imap/imapd.c

```diff
@@ -6875,7 +6875,9 @@ static void cmd_create(char *tag, char *name, struct dlist *extargs, int localon
         name[strlen(name)-1] = '\0';
     }
 
-    mbname_t *mbname = mbname_from_extname(name, &imapd_namespace, imapd_userid);
+    char *intname = mboxname_from_external(name, &imapd_namespace, imapd_userid);
+    mbname_t *mbname = mbname_from_intname(intname);
+    free(intname);
```

Comment:
wait, how can this help, when mboxname_from_external is just a wrapper around mbname_from_extname?

Reply:
hah, yeah.  So here's the deal:

extname: domain.com!user.foo parses to this:

```
username: NULL
domain: NULL
boxes: ["domain", "com!user", "foo"]
```

which converts back to the intname: "domain.com!user.foo".  If you round trip this a second time you get:

```
username: foo
domain: domain.com
boxes: []
```

BUT if you don't roundtrip through the intname, you wind up treating it as share namespace in the cmd_create code, but a user in the mboxlist_createmailbox code.

So this commit reverts to existing behaviour by taking the namespacelock with the second case as well, hence locking `foo@domain.com` rather than `""` (the shared namespace).

This is still somewhat bogus - we should just reject the bad extname, but I wanted to not change any existing behaviour with this commit.

Reply:
I see, so it relies on the behaviour that "make intname from an mbname with only boxes, no localpart or domain" is just concatenating the boxes back together with hiersep, but which accidentally produces a correct intname here, which can then be safely constructed into the mbname we actually want to use?

Yeah, that is still fairly bogus! But I agree, this PR isn't the place to fix it